// app/MailManager.js
// https://www.parse.com/docs/js/api/symbols/Mailgun.html
// https://www.npmjs.com/package/mailgun
// https://www.npmjs.com/package/mailgun-js
// http://blog.mailgun.com/how-to-send-transactional-emails-in-a-nodejs-app-using-the-mailgun-api/
var Mailgun = require("mailgun-js");
var config = require('app/config.json');
var request = require("request");

exports.mailSend=function(to,subj,body,context) {
  // doesn't check if to is valid. Will throw err in callback if invalid
  var mg = new Mailgun({apiKey: config.MAILGUN_KEY, domain: config.MAILGUN_DOMAIN});
  mg.messages().send({
    from: config.MAILGUN_FROM,
    to: to,
    subject: subj,
    html: body
  },
  function(err,body) {
    if(err) {
      context.fail("Forbidden to send email. "+err); // Invalid email or Did you update the mailgun parameters in config.php?
    } else {
      context.succeed("Queued. Thank you.");
    }
  });

};

module.exports.mailValidate = function(email, callback) {
// callback: function(err,result)
// https://github.com/dimoreira/mailgun-email-validation/blob/master/lib/mailgun-email-validation.js
// https://documentation.mailgun.com/api-email-validation.html#email-validation

  var config2 = {
    url: 'https://api.mailgun.net/v3/address/validate',
    username: 'api',
    key: config.MAILGUN_PUBLIC_KEY
  };

  var options = {
    url: config2.url,
    method: 'GET',
    qs: { address: email },
    auth: {
      username: config2.username,
      password: config2.key
    }
  };
  request(options, function(err, result) {
    if (err) callback(err, null);
    var response = JSON.parse(result.request.response.body);
    callback(null, response.is_valid);
  });
};
