// app/MailManager.js
// https://www.parse.com/docs/js/api/symbols/Mailgun.html
// https://www.npmjs.com/package/mailgun
// https://www.npmjs.com/package/mailgun-js
// http://blog.mailgun.com/how-to-send-transactional-emails-in-a-nodejs-app-using-the-mailgun-api/
var Mailgun = require("mailgun-js");
var MailgunValidation = require("mailgun-email-validation"); // https://www.npmjs.com/package/mailgun-email-validation
var config = require('app/config.json');


exports.mailSend=function(to,subj,body,context) {
  // doesn't check if to is valid. Will throw err in callback if invalid
  var mg = new Mailgun({apiKey: config.MAILGUN_KEY, domain: config.MAILGUN_DOMAIN});
  mg.messages().send({
    from: config.MAILGUN_FROM,
    to: to,
    subject: subj,
    html: body
  },
  function(err,body) {
    if(err) {
      context.fail("Forbidden to send email. "+err); // Invalid email or Did you update the mailgun parameters in config.php?
    } else {
      context.succeed("Queued. Thank you.");
    }
  });

}

exports.mailValidate=function(email,context) {
  MailgunValidation.check(email, function(err, valid) {
    if(err) context.fail(err); else context.succeed(valid);
  });
}
