// app/Validator.js

exports.mapIsf = function() {
return {	'B':'2:B',
		'G':'3:G',
		//'not coded':'1:not coded',
		'R':'4:R',
		'Z':'5:Z',
		'S':'6:S',
		'T':'7:T',
		'D':'8:D',
		'J':'9:J',
		'M':'10:M',
		'N':'11:N',
		'O':'12:O'
		//'40:Emirates', '14:Qatar', '39:Lathkia', '44:California',
		//'45:Foreign Plate', '19:Saudi Arabia', '20:Ruins',
		//'21:Without Restrictions', '22:Homos', '23:Idlib',
		//'24:Damascus', '25:Kuwait', '27:Sharjah', '28:UN',
		//'46:Deir al-zour', '41:Tartous', '34:Abou Dhabi', '49:Turkey',
		//'50:Raqaa', '51:Daraa', '55:Iraq', '26:Dubai', '30:Switzerland',
		//'31:Halab', '32:Hama', '33:Jordan', '35:BLP', '36:Syria',
		//'37:Bahrain', '38:Sweida', '43:Hasakeh', '47:Oman', '48:Tried'
};
};

exports.mapPml=function() {
return {'B':"B/ب",
	'G':"G/ج",
	'D':"D/د",
	'R':"R/ر",
	'Z':"Z/ز",
	'S':"S/ص",
	'T':"T/ط",
	'J':"J/ع",
	'M':"M/م",
	'N':"N/ن",
	'O':"O/و"//,
//	'C':"C/ق"
};

};

exports.validArea=function(a) {
return this.mapIsf().hasOwnProperty(a) && this.mapPml().hasOwnProperty(a);
};

exports.validDawlatiMechanique=function(fat31,t,tn) {
	fat32 = [];
	for(i=0;i<fat31.length;i++) fat32.push(fat31[i].nodeValue);
	return fat32.indexOf(t)!=-1; // return("Invalid "+tn+" '"+t+"'. Please use one of: "+fat32.join(', '));
};

exports.validNumber=function(n) { return !isNaN(n); };

exports.validContext = function(context) {
  // context should have succeed and fail functions
  if(!context.hasOwnProperty("succeed")) { throw("Context should have succeed function"); return false; }
  if(!context.hasOwnProperty("fail")) { throw("Context should have fail function"); return false; }
  // http://stackoverflow.com/a/6000009
  if(typeof(context.succeed) !== "function") { throw("Context succeed is not a function"); return false; }
  if(typeof(context.fail) !== "function") { throw("Context fail is not a function"); return false; }
  return true;
};
